Initial Commit For api-01-local-function Objective

MarvelProgramming opened the pull request:
For issue #35

Commit:
Initial Commit For api-01-local-function Objective

MarvelProgramming commented:
As it stands, each of the objectives (i.e. this one, #57, #58, #59, and #60) are defining a `DIVINATION_API_ENDPOINT` constant in their validator. I feel as though adding this to the `scripts/config` and referencing that instead would be better overall. Open to thoughts here!

Commit:
Updated Objective

-Added first pass at a description
-Fixed bug where one of the examples would throw an error because of the replacement string
-Updated validator to properly accept all of the necessary inputs for the user's function

Commit:
Added Content To Walkthrough.md

rmkubik commented:
I think we'd want these objectives in the `lovelace_tower` level from the feature branch:
https://github.com/TwilioQuest/twilioquest-api-academy/tree/milestone-1-aa3/levels/lovelace_tower

rmkubik commented:
General outline looks good! Using a var in scripts/config makes sense. Guessing that's in the other objective drafts though, not this one? Since it doesn't hit the backend

MarvelProgramming commented:
> I think we'd want these objectives in the `lovelace_tower` level from the feature branch: https://github.com/TwilioQuest/twilioquest-api-academy/tree/milestone-1-aa3/levels/lovelace_tower

My current plan is to merge all `milestone-1-aa3` into each of the objective branches and change the PR base. Thought about rebasing, but that doesn't feel like the best way to go because the branches are already here on the remote

\*UPDATE\*: All done!

> Guessing that's in the other objective drafts though, not this one? Since it doesn't hit the backend

Exactly!

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 35-api-01-local-function-local-api-function-call

Commit:
Moved Objective Into lovelace_tower

Commit:
Moved Objective level.json Entry To lovelace_tower

MarvelProgramming commented on levels/lovelace_tower/objectives/api-01-local-function/walkthrough.md:
```diff
@@ -0,0 +1,4 @@
+# Helpful Resources
+
+- [MDN - String replaceAll method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll)
```
Not sure what (if anything) more we wanted to do with the walkthrough for this objective. We're not really introducing anything new to the player, besides maybe the `replaceAll` method. Welcome to any ideas!

joenash replied:
Maybe suggest they revisit the Turing initiation in part 1? That covers the local API topic

MarvelProgramming replied:
~~Is there a specific part of the Turing initiation you're referring to? I was having trouble finding this while playing through the scenario. There is the tome at the beginning of the `lovelace_tower/default` map that links to resources for APIs, if that's what you are pointing to!~~

\*UPDATE\*: Scratch that, had a bit of a brain fart

nmadadair replied:
Tomb -> tome

joenash commented on levels/lovelace_tower/objectives/api-01-local-function/description.md:
```diff
@@ -0,0 +1,29 @@
+</ul>
+</div>
+
+Before we jump into working with Web APIs let's try interacting with a function locally. For this, we'll use the `String.replaceAll` method to replace all instances of a particular string with something else.
```
```suggestion
Before we jump into working with Web APIs let's remind ourselves about local APIs provided by libraries, as functions and objects. For this, we'll use the `String.replaceAll` method to replace all instances of a particular string with something else.
```

Minor suggestion to make the "libraries expose APIs too" link explicit

MarvelProgramming replied:
I like this! It really hits home on the idea that APIs extend beyond the Web!

joenash commented on levels/lovelace_tower/objectives/api-01-local-function/description.md:
```diff
@@ -0,0 +1,29 @@
+
+would print `one zero one zero one zero` to the console.
+
+Create a function called `wordSwapper` that takes 3 string arguments -- `sourceString`, `targetWord`, and `replacementWord`. Then replace all instances of `targetWord` in the `sourceString` with `replacementWord` and return the result.
```
`that takes 3 string arguments -- `sourceString`, `targetWord`, and `replacementWord`. `
Would these 3 elements be better as a bulleted list?

MarvelProgramming replied:
I think so!

joenash reviewed (approved):
A couple of minor suggestions but nothing breaking! Looking great

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-01-local-function/description.md

Updated Description verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 35-api-01-local-function-local-api-function-call

Commit:
Updated Objective

-Updated verbiage in description, objective.json, example, and walkthrough
-Updated function name to "swapStrings"
-Added bulleted list to description for function arguments

MarvelProgramming commented:
From what I recall, we wanted to tie this objective into the corrupted inscriptions the player later stumbles upon in the library. How did we want to tackle this without spoiling the story parts of the library/api-05? This may not be something we're worried about, but thought I'd check in before changing things further @rmkubik @nmadadair

Commit:
Updated Walkthrough

rmkubik reviewed (approved):
We talked about swapping the replacement strings to be more thematic. Since we're passing in the strings during validation though, I'm less worried about it.

nmadadair commented:
In terms of the narrativization of this objective, we could add something pretty simple at the very beginning of the description that acknowledges that this objective is the very first challenge of the House Gauntlet. Something along the lines of: 

> _Welcome, Gauntleter, to the first of many challenges in the Arcane Academy's annual House Gauntlet. As you progress through Lovelace Tower, you will be tested on your ability to apply your knowledge of APIs. For this first challenge, the House Gauntlet asks you to explore how a local function API can manipulate the written word, as befits House Lovelace._

My idea here is just to call out that we're able to use APIs to change text and language, and thematically House Lovelace is the house of documentation and books and thus text/language, etc. To your question above, @marvjohnson, about foreshadowing the Fredric corruption objective without giving it away, we could add a note at the end of the description  about how: 

> _Local function APIs like `swapStrings()`, though simple, have the power to alter not just individual sentences, but entire tomes. A Gauntleter might imagine the chaos if one were to swap every " " in a House Lovelace volume for a "" -- how easily important documents could be corrupted -- or repaired -- through such simple APIs!_

Commit:
Updated Objective

-Updated description and examples to be more thematic

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 35-api-01-local-function-local-api-function-call

MarvelProgramming commented:
> We talked about swapping the replacement strings to be more thematic. Since we're passing in the strings during validation though, I'm less worried about it.

Went in and made some changes such that the examples mesh with the story a bit better! Also @nmadadair, I went in and implemented the changes you suggested!
